File: Dopamine.FullPlayerModule/ViewModels/StatusViewModel.vb
Imports Dopamine.Common.Services.Indexing
Imports Dopamine.Common.Services.Update
Imports Dopamine.Core
Imports Dopamine.Core.Base
Imports Dopamine.Core.IO
Imports Dopamine.Core.Logging
Imports Dopamine.Core.Settings
Imports Dopamine.Core.Utils
Imports Microsoft.Practices.Prism.Commands
Imports Microsoft.Practices.Prism.Mvvm

Namespace ViewModels
    Public Class StatusViewModel
        Inherits BindableBase

#Region "Variables"
        ' Services
        Private mIndexingService As IIndexingService
        Private mUpdateService As IUpdateService

        ' Indexing
        Private mIsIndexing As Boolean
        Private mIndexingProgress As String

        Private mIsIndexerScanningFiles As Boolean
        Private mIsIndexerRemovingSongs As Boolean
        Private mIsIndexerAddingSongs As Boolean
        Private mIsIndexerUpdatingSongs As Boolean
        Private mIsIndexerUpdatingArtwork As Boolean

        ' Update status
        Private mIsUpdateAvailable As Boolean
        Private mVersionInfo As VersionInfo
        Private mDestinationPath As String
        Private mUpdateToolTip As String
        Private mIsUpdateStatusHiddenByUser As Boolean
        Private mShowInstallUpdateButton As Boolean
#End Region

#Region "Properties"
        ' Status bar
        Public ReadOnly Property IsStatusBarVisible() As Boolean
            Get
                Return mIsIndexing Or mIsUpdateAvailable
            End Get
        End Property

        ' Indexing
        Public Property IsIndexing() As Boolean
            Get
                Return mIsIndexing
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsIndexing, value)
                OnPropertyChanged(Function() Me.IsStatusBarVisible)
            End Set
        End Property

        Public Property IndexingProgress() As String
            Get
                Return mIndexingProgress
            End Get
            Set(ByVal value As String)
                SetProperty(Of String)(mIndexingProgress, value)
            End Set
        End Property

        Public Property IsIndexerRemovingSongs() As Boolean
            Get
                Return mIsIndexerRemovingSongs
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsIndexerRemovingSongs, value)
            End Set
        End Property

        Public Property IsIndexerAddingSongs() As Boolean
            Get
                Return mIsIndexerAddingSongs
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsIndexerAddingSongs, value)
            End Set
        End Property

        Public Property IsIndexerUpdatingSongs() As Boolean
            Get
                Return mIsIndexerUpdatingSongs
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsIndexerUpdatingSongs, value)
            End Set
        End Property

        Public Property IsIndexerUpdatingArtwork() As Boolean
            Get
                Return mIsIndexerUpdatingArtwork
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsIndexerUpdatingArtwork, value)
            End Set
        End Property

        ' Update status
        Public Property IsUpdateAvailable() As Boolean
            Get
                Return mIsUpdateAvailable
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsUpdateAvailable, value)
                OnPropertyChanged(Function() Me.IsStatusBarVisible)
            End Set
        End Property

        Public Property VersionInfo() As VersionInfo
            Get
                Return mVersionInfo
            End Get
            Set(ByVal value As VersionInfo)
                SetProperty(Of VersionInfo)(mVersionInfo, value)
            End Set
        End Property

        Public Property UpdateToolTip() As String
            Get
                Return mUpdateToolTip
            End Get
            Set(ByVal value As String)
                SetProperty(Of String)(mUpdateToolTip, value)
            End Set
        End Property

        Public ReadOnly Property ShowInstallUpdateButton() As Boolean
            Get
                Return Not String.IsNullOrEmpty(mDestinationPath)
            End Get
        End Property
#End Region

#Region "Commands"
        Public Property DownloadOrInstallUpdateCommand As DelegateCommand
        Public Property HideUpdateStatusCommand As DelegateCommand
#End Region

#Region "Construction"
        Public Sub New(iUpdateService As IUpdateService, iIndexingService As IIndexingService)

            mIndexingService = iIndexingService
            mUpdateService = iUpdateService

            Me.DownloadOrInstallUpdateCommand = New DelegateCommand(AddressOf Me.DownloadOrInstallUpdate)
            Me.HideUpdateStatusCommand = New DelegateCommand(Sub()
                                                                 mIsUpdateStatusHiddenByUser = True
                                                                 Me.IsUpdateAvailable = False
                                                             End Sub)

            AddHandler mIndexingService.IndexingStatusChanged, Async Sub(iIndexingStatusEventArgs) Await IndexingService_IndexingStatusChangedAsync(iIndexingStatusEventArgs)
            AddHandler mIndexingService.IndexingStopped, AddressOf IndexingService_IndexingStopped

            AddHandler mUpdateService.NewDownloadedVersionAvailable, AddressOf NewVersionAvailableHandler
            AddHandler mUpdateService.NewOnlineVersionAvailable, AddressOf NewVersionAvailableHandler
            AddHandler mUpdateService.NoNewVersionAvailable, AddressOf NoNewVersionAvailableHandler
            AddHandler mUpdateService.UpdateCheckDisabled, Sub() Me.IsUpdateAvailable = False

            If XmlSettingsClient.Instance.Get(Of Boolean)("Updates", "CheckForUpdates") Then
                mUpdateService.EnableUpdateCheck()
            End If

            ' Initial status
            Me.IsIndexing = False
            Me.IsUpdateAvailable = False
        End Sub
#End Region

#Region "Event Handlers"
        Private Async Function IndexingService_IndexingStatusChangedAsync(iIndexingStatusEventArgs As IndexingStatusEventArgs) As Task

            Await Task.Run(Sub()
                               Me.IsIndexing = mIndexingService.IsIndexing

                               If Me.IsIndexing Then

                                   mUpdateService.DisableUpdateCheck()
                                   Me.IsUpdateAvailable = False

                                   Select Case iIndexingStatusEventArgs.IndexingAction
                                       Case IndexingAction.RemoveTracks
                                           Me.IsIndexerRemovingSongs = True
                                           Me.IsIndexerAddingSongs = False
                                           Me.IsIndexerUpdatingSongs = False
                                           Me.IsIndexerUpdatingArtwork = False
                                           Me.IndexingProgress = String.Empty
                                       Case IndexingAction.AddTracks
                                           Me.IsIndexerRemovingSongs = False
                                           Me.IsIndexerAddingSongs = True
                                           Me.IsIndexerUpdatingSongs = False
                                           Me.IsIndexerUpdatingArtwork = False
                                           Me.IndexingProgress = "(" & Replace(Replace(ResourceUtils.GetStringResource("Language_Current_Of_Total"), "%current%", iIndexingStatusEventArgs.ProgressCurrent.ToString), "%total%", iIndexingStatusEventArgs.ProgressTotal.ToString) & ")"
                                       Case IndexingAction.UpdateTracks
                                           Me.IsIndexerRemovingSongs = False
                                           Me.IsIndexerAddingSongs = False
                                           Me.IsIndexerUpdatingSongs = True
                                           Me.IsIndexerUpdatingArtwork = False
                                           Me.IndexingProgress = "(" & Replace(Replace(ResourceUtils.GetStringResource("Language_Current_Of_Total"), "%current%", iIndexingStatusEventArgs.ProgressCurrent.ToString), "%total%", iIndexingStatusEventArgs.ProgressTotal.ToString) & ")"
                                       Case IndexingAction.UpdateArtwork
                                           Me.IsIndexerRemovingSongs = False
                                           Me.IsIndexerAddingSongs = False
                                           Me.IsIndexerUpdatingSongs = False
                                           Me.IsIndexerUpdatingArtwork = True
                                           Me.IndexingProgress = String.Empty
                                       Case Else
                                           ' Never happens
                                   End Select
                               Else
                                   Me.IndexingProgress = String.Empty

                                   If XmlSettingsClient.Instance.Get(Of Boolean)("Updates", "CheckForUpdates") Then
                                       mUpdateService.EnableUpdateCheck()
                                   End If
                               End If
                           End Sub)
        End Function

        Private Sub IndexingService_IndexingStopped(sender As Object, e As EventArgs)

            If Me.IsIndexing Then
                Me.IsIndexing = False
                Me.IndexingProgress = String.Empty

                If XmlSettingsClient.Instance.Get(Of Boolean)("Updates", "CheckForUpdates") Then
                    mUpdateService.EnableUpdateCheck()
                End If
            End If
        End Sub
#End Region

#Region "Private"
        Private Overloads Sub NewVersionAvailableHandler(iVersionInfo As VersionInfo)
            Me.NewVersionAvailableHandler(iVersionInfo, String.Empty)
        End Sub

        Private Overloads Sub NewVersionAvailableHandler(iVersionInfo As VersionInfo, iDestinationPath As String)

            If Not mIsUpdateStatusHiddenByUser AndAlso Not Me.IsIndexing Then
                Me.VersionInfo = iVersionInfo
                Me.IsUpdateAvailable = True

                mDestinationPath = iDestinationPath
                OnPropertyChanged(Function() Me.ShowInstallUpdateButton)

                If Not String.IsNullOrEmpty(iDestinationPath) Then
                    Me.UpdateToolTip = ResourceUtils.GetStringResource("Language_Click_Here_To_Install")
                Else
                    Me.UpdateToolTip = ResourceUtils.GetStringResource("Language_Click_Here_To_Download")
                End If
            End If
        End Sub

        Private Overloads Sub NoNewVersionAvailableHandler(iVersionInfo As VersionInfo)

            Me.IsUpdateAvailable = False
        End Sub

        Private Sub DownloadOrInstallUpdate()

            If Not String.IsNullOrEmpty(mDestinationPath) Then
                Try
                    ' A file was downloaded. Start the installer.
                    Dim msiFileInfo As System.IO.FileInfo = New System.IO.DirectoryInfo(mDestinationPath).GetFiles("*" & PackagingInformation.GetInstallablePackageFileExtesion).First
                    Process.Start(msiFileInfo.FullName)
                Catch ex As Exception
                    LogClient.Instance.Logger.Error("Could not start the MSI installer. Download link was opened instead. Exception: {0}", ex.Message)
                    Me.OpenDownloadLink()
                End Try
            Else
                ' Nothing was downloaded, forward to the download site.
                Me.OpenDownloadLink()
            End If
        End Sub

        Private Sub OpenDownloadLink()
            Try
                Dim downloadLink As String = String.Empty

                If Me.VersionInfo.Configuration = Configuration.Debug Then
                    downloadLink = UpdateInformation.PreReleaseDownloadLink
                ElseIf Me.VersionInfo.Configuration = Configuration.Release Then
                    downloadLink = UpdateInformation.ReleaseDownloadLink
                End If

                Actions.TryOpenLink(downloadLink)
            Catch ex As Exception
                LogClient.Instance.Logger.Error("Could not open the download link. Exception: {0}", ex.Message)
            End Try
        End Sub
#End Region
    End Class
End Namespace

File: Dopamine.FullPlayerModule/ViewModels/StatusViewModel.cs
using Digimezzo.Utilities.IO;
using Digimezzo.Utilities.Settings;
using Digimezzo.Utilities.Utils;
using Dopamine.Common.Services.Indexing;
using Dopamine.Common.Services.Update;
using Dopamine.Common.Base;
using Digimezzo.Utilities.Log;
using Prism.Commands;
using Prism.Mvvm;
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Digimezzo.Utilities.Packaging;

namespace Dopamine.FullPlayerModule.ViewModels
{
    public class StatusViewModel : BindableBase
    {
        #region Variables
        // Services
        private IIndexingService indexingService;
        private IUpdateService updateService;

        // Indexing
        private bool isIndexing;
        private string indexingProgress;
        private bool isIndexerRemovingSongs;
        private bool isIndexerAddingSongs;
        private bool isIndexerUpdatingSongs;
        private bool isIndexerUpdatingArtwork;

        // Update status
        private bool isUpdateAvailable;
        private Package package;
        private string destinationPath;
        private string updateToolTip;
        private bool isUpdateStatusHiddenByUser;
        #endregion

        #region Properties
        // Status bar
        public bool IsStatusBarVisible
        {
            get { return this.isIndexing | this.isUpdateAvailable; }
        }

        // Indexing
        public bool IsIndexing
        {
            get { return this.isIndexing; }
            set
            {
                SetProperty<bool>(ref this.isIndexing, value);
                OnPropertyChanged(() => this.IsStatusBarVisible);
            }
        }

        public string IndexingProgress
        {
            get { return this.indexingProgress; }
            set { SetProperty<string>(ref this.indexingProgress, value); }
        }

        public bool IsIndexerRemovingSongs
        {
            get { return this.isIndexerRemovingSongs; }
            set { SetProperty<bool>(ref this.isIndexerRemovingSongs, value); }
        }

        public bool IsIndexerAddingSongs
        {
            get { return this.isIndexerAddingSongs; }
            set { SetProperty<bool>(ref this.isIndexerAddingSongs, value); }
        }

        public bool IsIndexerUpdatingSongs
        {
            get { return this.isIndexerUpdatingSongs; }
            set { SetProperty<bool>(ref this.isIndexerUpdatingSongs, value); }
        }

        public bool IsIndexerUpdatingArtwork
        {
            get { return this.isIndexerUpdatingArtwork; }
            set { SetProperty<bool>(ref this.isIndexerUpdatingArtwork, value); }
        }

        // Update status
        public bool IsUpdateAvailable
        {
            get { return this.isUpdateAvailable; }
            set
            {
                SetProperty<bool>(ref this.isUpdateAvailable, value);
                OnPropertyChanged(() => this.IsStatusBarVisible);
            }
        }

        public Package Package
        {
            get { return this.package; }
            set { SetProperty<Package>(ref this.package, value); }
        }

        public string UpdateToolTip
        {
            get { return this.updateToolTip; }
            set { SetProperty<string>(ref this.updateToolTip, value); }
        }

        public bool ShowInstallUpdateButton
        {
            get { return !string.IsNullOrEmpty(this.destinationPath); }
        }
        #endregion

        #region Commands
        public DelegateCommand DownloadOrInstallUpdateCommand { get; set; }
        public DelegateCommand HideUpdateStatusCommand { get; set; }
        #endregion

        #region Construction
        public StatusViewModel(IUpdateService updateService, IIndexingService indexingService)
        {
            this.updateService = updateService;
            this.indexingService = indexingService;

            this.DownloadOrInstallUpdateCommand = new DelegateCommand(this.DownloadOrInstallUpdate);
            this.HideUpdateStatusCommand = new DelegateCommand(() =>
            {
                this.isUpdateStatusHiddenByUser = true;
                this.IsUpdateAvailable = false;
            });

            this.indexingService.IndexingStatusChanged += async(indexingStatusEventArgs) => await IndexingService_IndexingStatusChangedAsync(indexingStatusEventArgs);
            this.indexingService.IndexingStopped += IndexingService_IndexingStopped;

            this.updateService.NewDownloadedVersionAvailable += NewVersionAvailableHandler;
            this.updateService.NewOnlineVersionAvailable += NewVersionAvailableHandler;
            this.updateService.NoNewVersionAvailable += NoNewVersionAvailableHandler;
            this.updateService.UpdateCheckDisabled += (_,__) => this.IsUpdateAvailable = false;

            if (SettingsClient.Get<bool>("Updates", "CheckForUpdates"))
            {
                this.updateService.EnableUpdateCheck();
            }

            // Initial status
            this.IsIndexing = false;
            this.IsUpdateAvailable = false;
        }
        #endregion

        #region Event Handlers
        private async Task IndexingService_IndexingStatusChangedAsync(IndexingStatusEventArgs indexingStatusEventArgs)
        {
            await Task.Run(() =>
            {
                this.IsIndexing = this.indexingService.IsIndexing;

                if (this.IsIndexing)
                {
                    this.updateService.DisableUpdateCheck();
                    this.IsUpdateAvailable = false;

                    switch (indexingStatusEventArgs.IndexingAction)
                    {
                        case IndexingAction.RemoveTracks:
                            this.IsIndexerRemovingSongs = true;
                            this.IsIndexerAddingSongs = false;
                            this.IsIndexerUpdatingSongs = false;
                            this.IsIndexerUpdatingArtwork = false;
                            this.IndexingProgress = string.Empty;
                            break;
                        case IndexingAction.AddTracks:
                            this.IsIndexerRemovingSongs = false;
                            this.IsIndexerAddingSongs = true;
                            this.IsIndexerUpdatingSongs = false;
                            this.IsIndexerUpdatingArtwork = false;
                            this.IndexingProgress = this.FillProgress(indexingStatusEventArgs.ProgressCurrent.ToString(), indexingStatusEventArgs.ProgressTotal.ToString());
                            break;
                        case IndexingAction.UpdateTracks:
                            this.IsIndexerRemovingSongs = false;
                            this.IsIndexerAddingSongs = false;
                            this.IsIndexerUpdatingSongs = true;
                            this.IsIndexerUpdatingArtwork = false;
                            this.IndexingProgress = this.FillProgress(indexingStatusEventArgs.ProgressCurrent.ToString(), indexingStatusEventArgs.ProgressTotal.ToString());
                            break;
                        case IndexingAction.UpdateArtwork:
                            this.IsIndexerRemovingSongs = false;
                            this.IsIndexerAddingSongs = false;
                            this.IsIndexerUpdatingSongs = false;
                            this.IsIndexerUpdatingArtwork = true;
                            this.IndexingProgress = string.Empty;
                            break;
                        default:
                            break;
                            // Never happens
                    }
                }
                else
                {
                    this.IndexingProgress = string.Empty;

                    if (SettingsClient.Get<bool>("Updates", "CheckForUpdates"))
                    {
                        this.updateService.EnableUpdateCheck();
                    }
                }
            });
        }

        private string FillProgress(string currentProgres, string totalProgress)
        {
            string progress = string.Empty;

            progress = "(" + ResourceUtils.GetStringResource("Language_Current_Of_Total") + ")";
            progress = progress.Replace("%current%", currentProgres);
            progress = progress.Replace("%total%", totalProgress);

            return progress;
        }

        private void IndexingService_IndexingStopped(object sender, EventArgs e)
        {
            if (this.IsIndexing)
            {
                this.IsIndexing = false;
                this.IndexingProgress = string.Empty;

                if (SettingsClient.Get<bool>("Updates", "CheckForUpdates"))
                {
                    this.updateService.EnableUpdateCheck();
                }
            }
        }
        #endregion

        #region Private
        private void NewVersionAvailableHandler(Package package)
        {
            this.NewVersionAvailableHandler(package, string.Empty);
        }

        private void NewVersionAvailableHandler(Package package, string destinationPath)
        {
            if (!this.isUpdateStatusHiddenByUser && !this.IsIndexing)
            {
                this.Package = package;
                this.IsUpdateAvailable = true;

                this.destinationPath = destinationPath;
                OnPropertyChanged(() => this.ShowInstallUpdateButton);

                if (!string.IsNullOrEmpty(destinationPath))
                {
                    this.UpdateToolTip = ResourceUtils.GetStringResource("Language_Click_Here_To_Install");
                }
                else
                {
                    this.UpdateToolTip = ResourceUtils.GetStringResource("Language_Click_Here_To_Download");
                }
            }
        }

        private void NoNewVersionAvailableHandler(Package package)
        {
            this.IsUpdateAvailable = false;
        }

        private void DownloadOrInstallUpdate()
        {
            if (!string.IsNullOrEmpty(this.destinationPath))
            {
                try
                {
                    // A file was downloaded. Start the installer.
                    System.IO.FileInfo msiFileInfo = new System.IO.DirectoryInfo(this.destinationPath).GetFiles("*" + package.InstallableFileExtension).First();
                    Process.Start(msiFileInfo.FullName);
                }
                catch (Exception ex)
                {
                    LogClient.Error("Could not start the MSI installer. Download link was opened instead. Exception: {0}", ex.Message);
                    this.OpenDownloadLink();
                }
            }
            else
            {
                // Nothing was downloaded, forward to the download site.
                this.OpenDownloadLink();
            }
        }

        private void OpenDownloadLink()
        {
            try
            {
                string downloadLink = string.Empty;

                if (this.Package.Configuration == Configuration.Debug)
                {
                    downloadLink = UpdateInformation.PreReleaseDownloadLink;
                }
                else if (this.Package.Configuration == Configuration.Release)
                {
                    downloadLink = UpdateInformation.ReleaseDownloadLink;
                }

                Actions.TryOpenLink(downloadLink);
            }
            catch (Exception ex)
            {
                LogClient.Error("Could not open the download link. Exception: {0}", ex.Message);
            }
        }
        #endregion

    }
}
